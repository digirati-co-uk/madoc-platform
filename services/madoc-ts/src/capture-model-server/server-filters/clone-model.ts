import { generateId } from '../../frontend/shared/capture-models/helpers/generate-id';
import { traverseDocument } from '../../frontend/shared/capture-models/helpers/traverse-document';
import { CaptureModel, Target } from '../../frontend/shared/capture-models/types/capture-model';

export function cloneModel(model: CaptureModel, { target }: { target?: Target[] }): CaptureModel {
  const id = model.id;
  // @todo handle cloning with revisions
  const idMap: Record<string, string> = {};
  traverseDocument(model.document, {
    visitEntity(entity) {
      const newId = generateId();
      idMap[entity.id] = newId;
      entity.id = newId;
      entity.revision = undefined;
    },
    visitField(field) {
      const newId = generateId();
      idMap[field.id] = newId;
      field.id = newId;
      field.revision = undefined;
      field.revises = undefined;
    },
    visitSelector(selector) {
      const newId = generateId();
      idMap[selector.id] = newId;
      selector.id = newId;
      selector.revisionId = undefined;
    },
  });

  // Apply the revises field.
  traverseDocument(model.document, {
    visitEntity(entity) {
      if (entity.revises) {
        entity.revises = idMap[entity.revises] || undefined;
      }
    },
    visitField(field) {
      if (field.revises) {
        field.revises = idMap[field.revises] || undefined;
      }
    },
    visitSelector(selector) {
      if (selector.revises) {
        selector.revises = idMap[selector.revises] || undefined;
      }
    },
  });

  model.id = generateId();
  model.revisions = [];
  // @todo contributors.
  // model.contributors = creator ? { [creator.id]: creator } : {};
  model.derivedFrom = id;
  model.target = target;

  return model;
}
