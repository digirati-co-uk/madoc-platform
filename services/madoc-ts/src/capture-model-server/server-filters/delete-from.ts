import { CaptureModel } from '../../frontend/shared/capture-models/types/capture-model';

export function deleteFrom(id: string, property: string, parent: CaptureModel['document']) {
  if ((parent as any).properties[property]) {
    (parent as any).properties[property] = (parent as any).properties[property].filter((e: any) => e.id !== id);
  }
}
