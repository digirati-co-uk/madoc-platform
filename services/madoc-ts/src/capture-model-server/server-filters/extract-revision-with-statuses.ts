import { CaptureModel } from '../../frontend/shared/capture-models/types/capture-model';

export function extractRevisionWithStatuses(model: CaptureModel, statuses: string[]): string[] {
  const toReturn = [];
  if (!model.revisions) {
    return [];
  }

  for (const revision of model.revisions) {
    if (revision.status) {
      if (statuses.indexOf(revision.status) !== -1) {
        toReturn.push(revision.id);
      }
    }
  }

  return toReturn;
}
