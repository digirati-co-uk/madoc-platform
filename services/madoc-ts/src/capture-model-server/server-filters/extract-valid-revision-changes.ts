import { traverseDocument } from '../../frontend/shared/capture-models/helpers/traverse-document';
import { validateRevision } from '../../frontend/shared/capture-models/helpers/validate-revision';
import { CaptureModel } from '../../frontend/shared/capture-models/types/capture-model';
import { BaseField } from '../../frontend/shared/capture-models/types/field-types';
import { RevisionRequest } from '../../frontend/shared/capture-models/types/revision-request';
import { BaseSelector } from '../../frontend/shared/capture-models/types/selector-types';
import { getBefore } from './get-before';

export function extractValidRevisionChanges(
  captureModel: CaptureModel,
  req: RevisionRequest,
  {
    allowAnonymous,
    allowCanonicalChanges,
    allowCustomStructure,
  }: {
    allowAnonymous?: boolean;
    allowCanonicalChanges?: boolean;
    allowCustomStructure?: boolean;
  }
) {
  if (!req.document.immutable) {
    req.document.immutable = true;
  }

  const filteredDocument = validateRevision(req, captureModel, {
    allowAnonymous,
    allowCanonicalChanges,
    allowCustomStructure,
  });
  const allIds: string[] = [];
  const mutations: Record<
    string,
    Array<
      | {
          id: string;
          type: 'deletion';
        }
      | {
          id: string;
          type: 'entity';
          entity: CaptureModel['document'];
          term: string;
          shallow?: boolean;
          before: [string | null, number];
        }
      | {
          id: string;
          type: 'field';
          field: BaseField;
          term: string;
          before: [string | null, number];
        }
      | {
          id: string;
          type: 'selector';
          selector: BaseSelector;
          parentSelector?: BaseSelector;
        }
      | {
          id: string;
          type: 'selector-self';
          selector: BaseSelector;
        }
    >
  > = {};

  if (filteredDocument) {
    // @todo if `allowCanonicalChanges` is true, then this strategy probably will not work.
    //   When this is set we need to allow updating all properties of the included entities and fields
    //   But we can assume they might be incomplete (not all properties in the document for example).
    //   A better strategy might be to simply replace the values shallow-ly

    // 1st traverse the revision to get the fields that need added.
    traverseDocument(filteredDocument, {
      visitField(field, term, parent) {
        // Removing this immutability check. Not sure where it was required.
        if (allowCanonicalChanges || /*parent.immutable && */ field.revision === req.revision.id) {
          allIds.push(field.id);
          mutations[parent.id] = mutations[parent.id] ? mutations[parent.id] : [];
          mutations[parent.id].push({
            id: field.id,
            type: 'field',
            field,
            term,
            before: getBefore(field, term, parent),
          });
        }
      },
      visitEntity(entity, term, parent) {
        if (parent && term && (allowCanonicalChanges || (!entity.immutable && entity.revision === req.revision.id))) {
          allIds.push(entity.id);
          mutations[parent.id] = mutations[parent.id] ? mutations[parent.id] : [];
          mutations[parent.id].push({
            id: entity.id,
            type: 'entity',
            entity,
            term,
            shallow: allowCanonicalChanges,
            before: getBefore(entity, term, parent),
          });
        }
      },
      visitSelector(selector, parent, isRevision, parentSelector) {
        if (allowCanonicalChanges) {
          allIds.push(selector.id);
          mutations[selector.id] = mutations[selector.id] ? mutations[selector.id] : [];
          mutations[selector.id].push({
            id: selector.id,
            type: 'selector-self',
            selector,
          });
        } else if (isRevision && parentSelector && selector.revisionId && selector.revisionId === req.revision.id) {
          allIds.push(selector.id);
          mutations[parentSelector.id] = mutations[parentSelector.id] ? mutations[parentSelector.id] : [];
          mutations[parentSelector.id].push({ id: selector.id, type: 'selector', selector, parentSelector });
        }
      },
    });
  }

  // Optionally traverse to find valid fields to be removed.
  const requestedToDelete = req.revision.deletedFields || [];
  const validToDelete: string[] = [];

  traverseDocument(captureModel.document, {
    visitEntity(entity) {
      if (!entity.immutable && requestedToDelete.indexOf(entity.id) !== -1) {
        validToDelete.push(entity.id);
      }
      if (allIds.indexOf(entity.id) === -1 && entity.revision === req.revision.id) {
        // This is a deletion
        mutations[entity.id] = [{ type: 'deletion', id: entity.id }];
      }
    },
    visitField(field) {
      if (!field.immutable && requestedToDelete.indexOf(field.id) !== -1) {
        validToDelete.push(field.id);
      }
      if (allIds.indexOf(field.id) === -1 && field.revision === req.revision.id) {
        // This is a deletion
        mutations[field.id] = [{ type: 'deletion', id: field.id }];
      }
    },
    visitSelector(selector, _, isRevision) {
      if (selector && isRevision && allIds.indexOf(selector.id) === -1 && selector.revisionId === req.revision.id) {
        // This is a deletion
        mutations[selector.id] = [{ type: 'deletion', id: selector.id }];
      }
    },
  });

  req.revision.deletedFields = validToDelete;

  return {
    mutations,
    deletions: validToDelete,
  };
}
