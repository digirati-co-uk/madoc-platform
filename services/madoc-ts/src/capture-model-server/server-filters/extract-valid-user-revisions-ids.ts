import { CaptureModel } from '../../frontend/shared/capture-models/types/capture-model';

export function extractValidUserRevisionsIds(model: CaptureModel, userId: number) {
  const userUrn = `urn:madoc:user:${userId}`;
  const returnMap = {
    myRevisions: [] as string[],
    approvedRevisions: [] as string[],
    excludedRevisions: [] as string[],
  };

  if (!model.revisions) {
    return returnMap;
  }

  for (const revision of model.revisions) {
    let visible = false;
    if (revision.approved) {
      visible = true;
      returnMap.approvedRevisions.push(revision.id);
    }
    if (revision.authors && revision.authors.indexOf(userUrn) !== -1) {
      visible = true;
      returnMap.myRevisions.push(revision.id);
    }

    if (!visible) {
      returnMap.excludedRevisions.push(revision.id);
    }
  }

  return returnMap;
}
