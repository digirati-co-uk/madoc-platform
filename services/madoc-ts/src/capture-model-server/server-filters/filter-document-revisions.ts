import { traverseDocument } from '../../frontend/shared/capture-models/helpers/traverse-document';
import { CaptureModel } from '../../frontend/shared/capture-models/types/capture-model';
import { deleteFrom } from './delete-from';

export function filterDocumentRevisions(
  document: CaptureModel['document'],
  excludeRevisions: string[],
  onlyRevisionFields?: boolean
): CaptureModel['document'] {
  traverseDocument(document, {
    visitEntity(entity, property, parent) {
      if (property && parent) {
        const hasRevisions = entity?.temp && entity.temp.revisions && entity.temp.revisions.length;
        if (onlyRevisionFields && !entity.revision && !hasRevisions) {
          console.log('deleted', entity.id);
          deleteFrom(entity.id, property, parent);
        }

        if (entity.revision && excludeRevisions.indexOf(entity.revision) !== -1) {
          parent.temp = parent.temp || {};
          parent.temp.revisions = parent.temp.revisions || [];
          parent.temp.revisions.push(entity.revision);

          deleteFrom(entity.id, property, parent);
        }
      }
    },
    visitField(field, property, parent) {
      const selectorRevisions = field.selector?.revisedBy?.filter(
        r => r.revisionId && excludeRevisions.indexOf(r.revisionId) !== -1
      );
      if (onlyRevisionFields && !field.revision) {
        deleteFrom(field.id, property, parent);
      }

      if (field.revision && property && parent) {
        parent.temp = parent.temp || {};
        parent.temp.revisions = parent.temp.revisions || [];
        parent.temp.revisions.push(field.revision);
        if (excludeRevisions.indexOf(field.revision) !== -1) {
          deleteFrom(field.id, property, parent);
        }
      }
    },
    visitSelector(selector, parent, isRevision, parentSelector) {
      if (selector.revisionId && isRevision && parentSelector && parentSelector.revisedBy) {
        parentSelector.revisedBy = parentSelector.revisedBy.filter(s => {
          if (!s.revisionId) {
            return s;
          }
          return excludeRevisions.indexOf(s.revisionId) === -1;
        });
      }
    },
  });

  return document;
}
