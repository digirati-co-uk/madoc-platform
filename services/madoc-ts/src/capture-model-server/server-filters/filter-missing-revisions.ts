import { traverseDocument } from '../../frontend/shared/capture-models/helpers/traverse-document';
import { CaptureModel } from '../../frontend/shared/capture-models/types/capture-model';
import { deleteFrom } from './delete-from';

export function filterMissingRevisions(captureModel: CaptureModel) {
  const allRevisions = (captureModel.revisions || []).map(r => r.id);

  traverseDocument(captureModel.document, {
    beforeVisitEntity(entity, property, parent) {
      if (property && parent && entity.revision && allRevisions.indexOf(entity.revision) === -1) {
        deleteFrom(entity.id, property, parent);
      }
    },
    visitField(field, property, parent) {
      if (property && parent && field.revision && allRevisions.indexOf(field.revision) === -1) {
        deleteFrom(field.id, property, parent);
      }
    },
    visitSelector(selector, parent, isRevision, parentSelector) {
      if (selector.revisionId && isRevision && parentSelector && parentSelector.revisedBy) {
        parentSelector.revisedBy = parentSelector.revisedBy.filter(s => {
          if (!s.revisionId) {
            return s;
          }
          return allRevisions.indexOf(s.revisionId) !== -1;
        });
      }
    },
  });

  return captureModel;
}
