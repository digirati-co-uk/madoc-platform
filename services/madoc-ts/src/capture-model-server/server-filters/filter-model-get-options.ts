import { CaptureModel } from '../../frontend/shared/capture-models/types/capture-model';
import { CaptureModelGetOptions } from '../types';
import { extractRevisionWithStatuses } from './extract-revision-with-statuses';
import { extractValidUserRevisionsIds } from './extract-valid-user-revisions-ids';
import { filterDocumentRevisions } from './filter-document-revisions';
import { filterMissingRevisions } from './filter-missing-revisions';
import { removeFieldsDeletedByPublishedRevisions } from './remove-fields-deleted-by-published-revisions';

export function filterModelGetOptions(
  fullModel: CaptureModel,
  {
    onlyRevisionFields,
    // context,
    // includeCanonical,
    userId,
    revisionStatuses,
    revisionId,
    revisionStatus,
    showAllRevisions,
    showDeletedFields,
    filterDeletedRevisions = true,
  }: CaptureModelGetOptions
): CaptureModel {
  if (userId && !showAllRevisions) {
    const { excludedRevisions } = extractValidUserRevisionsIds(fullModel, userId);
    // This is a mutate-y function.
    filterDocumentRevisions(fullModel.document, excludedRevisions, onlyRevisionFields);
    fullModel.revisions = (fullModel.revisions || []).filter(r => excludedRevisions.indexOf(r.id) === -1);
  }

  revisionStatuses = revisionStatuses || (revisionStatus ? [revisionStatus] : undefined);

  if (revisionId) {
    const toExclude = (fullModel.revisions || []).filter(r => r.id !== revisionId).map(r => r.id);
    filterDocumentRevisions(fullModel.document, toExclude, onlyRevisionFields);
    fullModel.revisions = (fullModel.revisions || []).filter(r => toExclude.indexOf(r.id) === -1);
  } else if (revisionStatuses && revisionStatuses.length) {
    const revisionIds = extractRevisionWithStatuses(fullModel, revisionStatuses);
    const toExclude = (fullModel.revisions || []).filter(r => revisionIds.indexOf(r.id) === -1).map(r => r.id);
    filterDocumentRevisions(fullModel.document, toExclude, onlyRevisionFields);
    fullModel.revisions = (fullModel.revisions || []).filter(r => toExclude.indexOf(r.id) === -1);
  }

  if (!showDeletedFields) {
    fullModel = removeFieldsDeletedByPublishedRevisions(fullModel);
  }

  if (filterDeletedRevisions) {
    filterMissingRevisions(fullModel);
  }

  return fullModel;
}
