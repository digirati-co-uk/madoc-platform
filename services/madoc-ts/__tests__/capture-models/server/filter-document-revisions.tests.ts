import { filterDocumentRevisions } from '../../../src/capture-model-server/server-filters/filter-document-revisions';

describe('Filter document revisions', () => {
  test('It can filter out new revision', () => {
    const model: any = {
      id: '6017f74e-48c8-4c27-ad26-cc673b563659',
      structure: {
        id: '7b2a10da-587f-4f61-80b5-709effa2968f',
        type: 'choice',
        label: 'A project with boxes',
        items: [
          {
            id: '838088d6-9f53-4a67-9c57-307a7695ad51',
            type: 'model',
            label: 'Default',
            fields: [['boxes', ['description']], 'test'],
            modelRoot: ['boxes'],
          },
        ],
      },
      document: {
        id: 'bb5aa1ec-bdb3-4e99-a47d-12f9bbc01d52',
        type: 'entity',
        label: 'A project with boxes',
        properties: {
          boxes: [
            {
              id: '54bf43db-765b-43fa-a71f-a9084c1187c2',
              type: 'entity',
              label: 'boxes',
              revision: '5325ebc1-2331-4a0c-a77a-27c7413ef743',
              properties: {
                description: [
                  {
                    id: '3066d80b-6370-4e84-8f09-2d7eb3fbc379',
                    type: 'text-field',
                    label: 'Description',
                    value: 'test',
                    description: 'What is in your box',
                    revision: '5325ebc1-2331-4a0c-a77a-27c7413ef743',
                    minLines: 6,
                    multiline: true,
                  },
                ],
              },
              selector: {
                id: '388f0c61-e2bc-40c5-9c7c-dc4e5e1ffebe',
                type: 'box-selector',
                state: {
                  x: 860,
                  y: 616,
                  width: 480,
                  height: 408,
                },
              },
            },
            {
              id: 'feaa163d-0668-4557-9494-6aa80659e203',
              type: 'entity',
              label: 'boxes',
              properties: {
                description: [
                  {
                    id: 'b84f3423-a914-4b1b-bb7a-1248c093f334',
                    type: 'text-field',
                    label: 'Description',
                    value: '',
                    description: 'What is in your box',
                    minLines: 6,
                    multiline: true,
                  } as any,
                ],
              },
              selector: {
                id: '3e079573-4229-4886-8fcb-815f3d9ed729',
                type: 'box-selector',
                state: null,
              },
            },
          ],
          test: [
            {
              id: 'ce769b17-f9c2-4eb8-a3db-c4f13da5ce81',
              type: 'autocomplete-field',
              label: 'test',
              value: null,
              revises: '53d1e719-95c9-4be2-9889-3b947eb1e8a1',
              revision: '5325ebc1-2331-4a0c-a77a-27c7413ef743',
              preset: 'bentham',
            } as any,
            {
              id: '53d1e719-95c9-4be2-9889-3b947eb1e8a1',
              type: 'autocomplete-field',
              label: 'test',
              value: '',
              preset: 'bentham',
            },
          ],
        },
      },
      target: [
        {
          id: 'urn:madoc:collection:903',
          type: 'Collection',
        },
        {
          id: 'urn:madoc:manifest:904',
          type: 'Manifest',
        },
        {
          id: 'urn:madoc:canvas:942',
          type: 'Canvas',
        },
      ],
      profile: null,
      derivedFrom: '6c0c6464-b2e2-448c-8471-ddd5a3fa0d7a',
      revisions: [
        {
          structureId: '838088d6-9f53-4a67-9c57-307a7695ad51',
          approved: false,
          label: 'Default',
          id: '5325ebc1-2331-4a0c-a77a-27c7413ef743',
          fields: [['boxes', ['description']], 'test'],
          status: 'draft',
          revises: '838088d6-9f53-4a67-9c57-307a7695ad51',
          authors: ['urn:madoc:user:1'],
          deletedFields: null,
        },
      ],
      contributors: {
        'urn:madoc:user:1': {
          id: 'urn:madoc:user:1',
          type: 'Person',
          name: 'admin',
        },
      },
    };

    //
    const newDocument = filterDocumentRevisions(model.document, [], false);

    expect(newDocument).toMatchSnapshot();
  });
});
