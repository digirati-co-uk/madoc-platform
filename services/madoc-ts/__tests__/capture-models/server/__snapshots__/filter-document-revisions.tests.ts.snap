// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter document revisions It can filter out new revision 1`] = `
{
  "id": "bb5aa1ec-bdb3-4e99-a47d-12f9bbc01d52",
  "label": "A project with boxes",
  "properties": {
    "boxes": [
      {
        "id": "54bf43db-765b-43fa-a71f-a9084c1187c2",
        "label": "boxes",
        "properties": {
          "description": [
            {
              "description": "What is in your box",
              "id": "3066d80b-6370-4e84-8f09-2d7eb3fbc379",
              "label": "Description",
              "minLines": 6,
              "multiline": true,
              "revision": "5325ebc1-2331-4a0c-a77a-27c7413ef743",
              "type": "text-field",
              "value": "test",
            },
          ],
        },
        "revision": "5325ebc1-2331-4a0c-a77a-27c7413ef743",
        "selector": {
          "id": "388f0c61-e2bc-40c5-9c7c-dc4e5e1ffebe",
          "state": {
            "height": 408,
            "width": 480,
            "x": 860,
            "y": 616,
          },
          "type": "box-selector",
        },
        "type": "entity",
      },
      {
        "id": "feaa163d-0668-4557-9494-6aa80659e203",
        "label": "boxes",
        "properties": {
          "description": [
            {
              "description": "What is in your box",
              "id": "b84f3423-a914-4b1b-bb7a-1248c093f334",
              "label": "Description",
              "minLines": 6,
              "multiline": true,
              "type": "text-field",
              "value": "",
            },
          ],
        },
        "selector": {
          "id": "3e079573-4229-4886-8fcb-815f3d9ed729",
          "state": null,
          "type": "box-selector",
        },
        "type": "entity",
      },
    ],
    "test": [
      {
        "id": "ce769b17-f9c2-4eb8-a3db-c4f13da5ce81",
        "label": "test",
        "preset": "bentham",
        "revises": "53d1e719-95c9-4be2-9889-3b947eb1e8a1",
        "revision": "5325ebc1-2331-4a0c-a77a-27c7413ef743",
        "type": "autocomplete-field",
        "value": null,
      },
      {
        "id": "53d1e719-95c9-4be2-9889-3b947eb1e8a1",
        "label": "test",
        "preset": "bentham",
        "type": "autocomplete-field",
        "value": "",
      },
    ],
  },
  "type": "entity",
}
`;
